bundle/edu.gemini.pot/src/main/scala/edu/gemini/spModel/gemini/ghost/AsterismConverters.scala

```diff
@@ -0,0 +1,108 @@
+package edu.gemini.spModel.gemini.ghost
+
+import edu.gemini.shared.util.immutable.ImList
+import edu.gemini.spModel.core.{Coordinates, SiderealTarget}
+import edu.gemini.spModel.target.SPTarget
+import edu.gemini.spModel.target.env.{TargetEnvironment, UserTarget}
+
+import scalaz._
+import Scalaz._
+
+/** Allow conversions between target environments containing the different GHOST asterism types.
+  * Coordinates are immutable, but SPTargets are mutable, so we must clone them.
+  */
+object AsterismConverters {
+  import GhostAsterism._
+  import GhostAsterism.GhostStandardResTargets._
+
+  type BasePosition = Option[Coordinates]
+  type SkyPosition  = Option[Coordinates]
+
+  sealed trait AsterismConverter {
+    def name: String
+    def convert(env: TargetEnvironment): String \/ TargetEnvironment
+  }
+
+  sealed trait GhostAsterismConverter extends AsterismConverter {
+    override def convert(env: TargetEnvironment): String \/ TargetEnvironment = env.getAsterism match {
+        case StandardResolution(SingleTarget(t), b)     => creator(env, t,  None,    None,   b).right
+        case StandardResolution(DualTarget(t1, t2), b)  => creator(env, t1, t2.some, None,   b).right
+        case StandardResolution(TargetPlusSky(t, s), b) => creator(env, t,  None,    s.some, b).right
+        case StandardResolution(SkyPlusTarget(s, t), b) => creator(env, t,  None,    s.some, b).right
+        case HighResolution(t, s, b)                    => creator(env, t,  None,    s,      b).right
+        case _                                          => s"Could not convert to $name".left
```

Comment:
It would be nice if we could compute this in terms of a total match on `GhostAsterism` because we should be able to get exhaustive checking. I realize it's awkward because of the way the `TargetEnvironment` is structured. What if we did something like

```scala
env.getAsterism match {
  case ga: GhostAsterism =>
    ga match {
      case StandardResolution(...) => ...
      // ... etc., with no catch-all so we get exhaustive checking for this clause
    }
  case _ => // fail
```

We might also consider just using `Option` here since there's only one failure case and it always means the same thing. In general I don't like generating end-user messages this deep in the model. Debatable.

Reply:
Agreed. The error message, if we need it, could easily be constructed when the `Option` is accessed, and this will simplify things.

Reply:
Somehow I glazed over the earlier part of this comment in lieu of the `Option` part. Friday.
I'm fine with doing the exhaustive checking as you said: I'll make this change now and check it in momentarily.

Sorry about that!

Reply:
Incidentally, thanks for checking this PR. Much appreciated. It's a lot of crud that's probably overkill at this point, but will hopefully make things easier later on.